### PR TITLE
[LT] Move device lock in LazyGraphExecutor to a later place

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76974

Summary: cherry-picking https://github.com/pytorch/xla/pull/3457

### CODE REVIEW

**Review by wconstab — APPROVED** [review]
lgtm, thanks bin!

File: torch/csrc/lazy/core/lazy_graph_executor.cpp [hunk 1]
```diff
@@ -1096,12 +1098,30 @@ std::vector<BackendDataPtr> LazyGraphExecutor::GatherTensorsData(
   return result_tensors_data;
 }
 
+void LazyGraphExecutor::TensorCollectionBarrier(SyncTensorCollection* coll) {
+  static const std::string invalid_device(
+      "Unknown0"); /* Temp solution to idetify unassigned devices */
```
Comment (JackCaoG) [inline torch/csrc/lazy/core/lazy_graph_executor.cpp:1103]: just a note for myself to remove this after https://github.com/pytorch/pytorch/pull/76517/files merged and each backend implement the default device correctly.

**Review by JackCaoG — APPROVED** [review]
Thanks Bin!

**Pull request closed without merging**